### PR TITLE
Add SegmentedControlIOS

### LINKED ISSUE #534: UISegmentedControl support
Hey there, any thoughts on `UISegmentedControl` support?

I suppose it would look like this:

```
<SegmentedControlIOS
  values={["One", "Two"]}
  selectedValue={}
  momentary={false}
  onValueChange={() =>}
  tintColor={}
  // TODO: how to handle image items?
/>
```

or this better left to an add-on/plugin sort of thing?

### PR DESCRIPTION
**Pull request by clayallsopp**
Fixes #534:

screen shot 2015-03-31 at 7 52 10 pm

```jsx
<SegmentedControlIOS
  tintColor="#ff0000"
  values={["One", "Two", "Three", "Four"]}
  selectedValue={"One"}
  momentary={false}
  disabled={false}
  onValueChange={ (value) => console.log(value) } />
```

This only supports string-based segments, not images. Also doesn't support full customization (no separator images etc); I figure this is a good MVP to lock-down a basic API

I also included a snapshot test case, but the images keep coming out funky. When I look at the sim, I see that the text labels show up for the selected segment, but the snapshot keeps coming out with no text on those segments. I tried forcing a delay, but same result. Is that explainable?

Obviously happy to change anything about the API, code-style nitpicks, etc

### CODE REVIEW

File: React/Views/RCTSegmentedControl.m [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+//
+//  Created by Clay Allsopp on 3/31/15.
+//  Copyright (c) 2015 Facebook. All rights reserved.
+//
+
+#import "RCTSegmentedControl.h"
+#import "UIView+React.h"
+#import "RCTEventDispatcher.h"
+
+
```
Comment (a2) [inline React/Views/RCTSegmentedControl.m]: Nit: Remove extra line.

File: React/Views/RCTSegmentedControl.m [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+{
+  RCTEventDispatcher *_eventDispatcher;
+}
+
+- (id)initWithEventDispatcher:(RCTEventDispatcher *)eventDispatcher
+{
+  if (self = [super initWithFrame:CGRectZero]) {
+    _eventDispatcher = eventDispatcher;
+
+    [self addTarget:self
```
Comment (a2) [inline React/Views/RCTSegmentedControl.m]: Nit: Can you fix the whitespace alignment on this method?

File: React/Views/RCTSegmentedControl.m [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  return self;
+}
+
+
+- (void)setValuesAndSelectedValue:(NSDictionary *)valuesAndSelectedValue
+{
+  NSArray *values = valuesAndSelectedValue[@"values"];
+  [self removeAllSegments];
+  for (int i = 0; i < values.count; i++) {
```
Comment (a2) [inline React/Views/RCTSegmentedControl.m]: Fast enumeration with an external index variable (or `-enumerateObjectsUsingBlock:`) seems more appropriate.

Comment (a2) [inline React/Views/RCTSegmentedControl.m]: I would use a `NSUInteger` index to match `values.count` and `-objectAtIndex:`.

File: React/Views/RCTSegmentedControl.m [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    [self insertSegmentWithTitle:values[i] atIndex:i animated:NO];
+  }
+
+  NSString *selectedValue = valuesAndSelectedValue[@"selectedValue"];
+  NSInteger segmentIndex = -1;
+  for (int i = 0; i < self.numberOfSegments; i++) {
+    NSString *segmentValue = [self titleForSegmentAtIndex:i];
+    if ([segmentValue isEqualToString:selectedValue]) {
+      segmentIndex = i;
```
Comment (a2) [inline React/Views/RCTSegmentedControl.m]: You may want to `break` after this line.

File: React/Views/RCTSegmentedControl.m [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+- (void)setValuesAndSelectedValue:(NSDictionary *)valuesAndSelectedValue
+{
+  NSArray *values = valuesAndSelectedValue[@"values"];
+  [self removeAllSegments];
+  for (int i = 0; i < values.count; i++) {
+    [self insertSegmentWithTitle:values[i] atIndex:i animated:NO];
+  }
+
+  NSString *selectedValue = valuesAndSelectedValue[@"selectedValue"];
+  NSInteger segmentIndex = -1;
```
Comment (a2) [inline React/Views/RCTSegmentedControl.m]: `NSNotFound` is more often used for this purpose.

**Comment by a2** [comment]
Clay, this looks great. I only have a few nit requests but if you can make those edits, we can merge this straightaway. :smiley:

**Comment by ashwinb** [comment]
Hm, we should ideally just use `selectedIndex` in the backing implementation, and expose both `selectedValue` and `selectedIndex` to the user. Or just avoid `selectedValue` completely. It just seems like a bad idea that typos there will result in undefined behavior.

**Comment by a2** [comment]
@ashwinb Good idea. I like that better. The user should be able to set the `selectedIndex` and `values` separately.

**Comment by clayallsopp** [comment]
@a2 @ashwinb I went back and forth on this; I figured that, on the JS side, people would be doing the index -> value conversion on every callback to make their logic not dependent on ordering, so we might as well do it for them. I.e.:

```
_onSegmentValueChanged : (value) =>
  // this is resilient to ordering changes
  switch(value) {
    case "Profile":
      break;
  }
```

vs

```
_onSegmentValueChanged : (selectedIndex) =>
  value = this.props.segments[selectedIndex]
  // now it is resilient to order changes
  switch() {
  }
```

maybe a compromise is to 1) allow `selectedValue` and `selectedSegmentIndex` as props, and prefer `selectedIndex` if given 2) pass both `selectedSegmentIndex` and `selectedValue` in the callback `event`, for raw `onChange` listeners, but keep `onValueChange` returning the segment value for the case above

but my opinion isn't terribly strong, so will defer to your judgement

**Comment by ashwinb** [comment]
Yes, passing both `(index, value)` seems like a good idea in the callback, although I am fine even with just `(value)` as you have done. However, what I am opposed to is the `selectedValue` prop -- that is an unnecessary landmine of typos (unless, of course, you make it an enum, etc. if you are using something like `flowtype`). We should just avoid it if possible.

**Comment by clayallsopp** [comment]
I've updated the PR; check out the API examples

**Comment by vjeux** [comment]
It doesn't patch cleanly anymore, can you rebase? Sorry about this

**Comment by clayallsopp** [comment]
@vjeux done, np

**Comment by jozan** [comment]
What's the status of this?

File: React/Views/RCTSegmentedControlManager.m [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+- (UIView *)view
+{
+  return [[RCTSegmentedControl alloc] initWithEventDispatcher:self.bridge.eventDispatcher];
+}
+
+RCT_EXPORT_VIEW_PROPERTY(valuesAndSelectedSegmentIndex, NSDictionary);
+RCT_EXPORT_VIEW_PROPERTY(tintColor, UIColor);
+RCT_EXPORT_VIEW_PROPERTY(momentary, BOOL);
+RCT_EXPORT_VIEW_PROPERTY(enabled, BOOL);
+
```
Comment (jozan) [inline React/Views/RCTSegmentedControlManager.m]: Needs RCT_EXPORT_MODULE(); to work in current version.

**Comment by nicklockwood** [comment]
Hey, sorry about the radio silence on this. The internal diff sort of got lost in transmission for a while, however the good news is that this has now been merged and will ship with the next release.

**Comment by clayallsopp** [comment]
Lovely, thanks!

**Pull request closed without merging**